src/imp/libc/rand/syscalls.rs

```diff
@@ -7,12 +7,12 @@ use {super::super::c, super::super::conv::ret_ssize_t, crate::io, crate::rand::G
 
 #[cfg(target_os = "linux")]
 pub(crate) fn getrandom(buf: &mut [u8], flags: GetRandomFlags) -> io::Result<usize> {
-    let nread = unsafe {
-        ret_ssize_t(c::getrandom(
-            buf.as_mut_ptr().cast(),
-            buf.len(),
-            flags.bits(),
-        ))?
-    };
+    // `getrandom` wasn't supported in glibc until 2.25.
+    syscall! {
+        fn getrandom(buf: *mut c::c_void, buflen: c::size_t, flags: c::c_uint) via SYS_getrandom -> c::ssize_t
+    }
+
+    let nread =
+        unsafe { ret_ssize_t(getrandom(buf.as_mut_ptr().cast(), buf.len(), flags.bits()))? };
```

Comment:
Am I understanding this code correctly in that we now always invoke the system call?  Maybe I'm misreading the macro implementations, but I am not seeing the behavior of `syscall!` trying the weak pointer and falling back to the system call.

Reply:
Yes, that's right; this works by always using `libc::syscall`. Do you think it would be better to try the libc wrapper first and fall back?

Reply:
Of course in that in any system that is including non-Rust C/C++ code (like my case) unless one goes to extreme lengths on that side too, the resulting binary is  pulling in a hard dependency on specific glibc versions anyways.  So for me the dlsym dance is unnecessary.   I don't think this really matters, it almost certainly isn't worth trying to expand the matrix here to be (hard requirement, weak requirement, direct syscall) unless performance needs drive it.

Reply:
Yeah, and the `dlsym` code is just a workaround for `#[weak_extern]` not being stable yet in Rust. When they are, we'll switch `weak!` and `weakcall!` to use that instead of `dlsym`.

Thinking about this more though, I can see the argument that we should try to use the libc wrappers; that way people expecting to be able to replace them with `LD_PRELOAD`, or break on them in gdb, can do so. I've now added a `weak_or_syscall!` macro which implements this; does this look reasonable?